File: domain/src/main/java/com/arturogutierrez/openticator/domain/account/interactor/GetAccountsInteractor.java
package com.arturogutierrez.openticator.domain.account.interactor;

import com.arturogutierrez.openticator.domain.account.model.Account;
import com.arturogutierrez.openticator.domain.account.repository.AccountRepository;
import com.arturogutierrez.openticator.executor.PostExecutionThread;
import com.arturogutierrez.openticator.executor.ThreadExecutor;
import com.arturogutierrez.openticator.interactor.Interactor;
import java.util.List;
import rx.Observable;

public class GetAccountsInteractor extends Interactor<List<Account>> {

  private final AccountRepository accountRepository;

  public GetAccountsInteractor(AccountRepository accountRepository, ThreadExecutor threadExecutor,
      PostExecutionThread postExecutionThread) {
    super(threadExecutor, postExecutionThread);
    this.accountRepository = accountRepository;
  }

  @Override
  public Observable<List<Account>> createObservable() {
    return accountRepository.getAllAccounts();
  }
}

File: domain/src/main/kotlin/com/arturogutierrez/openticator/domain/account/interactor/GetAccountsInteractor.kt
package com.arturogutierrez.openticator.domain.account.interactor

import com.arturogutierrez.openticator.domain.account.interactor.GetAccountsInteractor.EmptyParams
import com.arturogutierrez.openticator.domain.account.model.Account
import com.arturogutierrez.openticator.domain.account.repository.AccountRepository
import com.arturogutierrez.openticator.executor.PostExecutionThread
import com.arturogutierrez.openticator.executor.ThreadExecutor
import com.arturogutierrez.openticator.interactor.FlowableUseCase
import io.reactivex.Flowable

class GetAccountsInteractor(private val accountRepository: AccountRepository,
                            threadExecutor: ThreadExecutor,
                            postExecutionThread: PostExecutionThread)
  : FlowableUseCase<List<Account>, EmptyParams>(threadExecutor, postExecutionThread) {

  override fun buildObservable(params: EmptyParams): Flowable<List<Account>> {
    return accountRepository.allAccounts
  }

  object EmptyParams
}
